From 1d56bfb436b008422b4a7d4d4e3180667130c840 Mon Sep 17 00:00:00 2001
From: Mahesh Bodapati <mbodapat@xilinx.com>
Date: Tue, 13 Sep 2022 11:49:11 +0530
Subject: [PATCH 21/53] [PATCH 21/53] [Patch, microblaze]: Correct the const
 high double  immediate value with this patch the loading of the DI mode
 immediate values  will be using REAL_VALUE_FROM_CONST_DOUBLE and
 REAL_VALUE_TO_TARGET_DOUBLE  functions, as CONST_DOUBLE_HIGH was returning
 the sign extension value even  of the unsigned long long constants also

Signed-off-by :Nagaraju Mekala <nmekala@xilix.com>
               Ajit Agarwal  <ajitkum@xilinx.com>
---
 gcc/config/microblaze/microblaze.cc               | 6 ++++--
 gcc/testsuite/gcc.target/microblaze/others/long.c | 9 +++++++++
 2 files changed, 13 insertions(+), 2 deletions(-)
 create mode 100644 gcc/testsuite/gcc.target/microblaze/others/long.c

diff --git a/gcc/config/microblaze/microblaze.cc b/gcc/config/microblaze/microblaze.cc
index bd394c411b8..7c648cda1b2 100644
--- a/gcc/config/microblaze/microblaze.cc
+++ b/gcc/config/microblaze/microblaze.cc
@@ -2453,14 +2453,16 @@ print_operand (FILE * file, rtx op, int letter)
   else if (letter == 'h' || letter == 'j')
     {
       long val[2];
+      long l[2];
       if (code == CONST_DOUBLE)
 	{
 	  if (GET_MODE (op) == DFmode)
 	    REAL_VALUE_TO_TARGET_DOUBLE (*CONST_DOUBLE_REAL_VALUE (op), val);
 	  else
 	    {
-	      val[0] = CONST_DOUBLE_HIGH (op);
-	      val[1] = CONST_DOUBLE_LOW (op);
+	      REAL_VALUE_TO_TARGET_DOUBLE (*CONST_DOUBLE_REAL_VALUE (op), l);
+	      val[1] = l[WORDS_BIG_ENDIAN == 0];
+	      val[0] = l[WORDS_BIG_ENDIAN != 0];
 	    }
 	}
       else if (code == CONST_INT)
diff --git a/gcc/testsuite/gcc.target/microblaze/others/long.c b/gcc/testsuite/gcc.target/microblaze/others/long.c
new file mode 100644
index 00000000000..b6b55d5ad65
--- /dev/null
+++ b/gcc/testsuite/gcc.target/microblaze/others/long.c
@@ -0,0 +1,9 @@
+#define BASEADDR 0xF0000000ULL
+int main ()
+{
+  unsigned long long start;
+  start = (unsigned long long) BASEADDR;
+  return 0;
+}
+/* { dg-final { scan-assembler "addik\tr(\[0-9]\|\[1-2]\[0-9]\|3\[0-1]),r0,0x00000000" } } */
+/* { dg-final { scan-assembler "addik\tr(\[0-9]\|\[1-2]\[0-9]\|3\[0-1]),r0,0xf0000000" } } */
-- 
2.37.1 (Apple Git-137.1)

